Add retention subsystem

nordicjm opened the pull request:
Adds a retention system which builds on top of retained memory drivers to allow partitioning of areas and data integrity with magic header prefixes and checksum of stored data.

nordicjm committed:
retention: Add retention system

Adds a retention system which builds on top of retained_mem
drivers to allow partitioning of areas and data integrity with
magic header prefixes and checksum of stored data.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
MAINTAINERS: Add Retention section and assign to self

Adds an entry for the retention subsystem and assign myself to it.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
retention: Add boot mode system

Adds a boot mode system which allows for redirecting the boot
target of a device depending upon the state of a retained value.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
dfu: mcuboot_shell: Add command to enter serial recovery

Adds a command which can be used to enter mcuboot's serial recovery
mode when the retention subsystem is enabled.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
soc: arm: nordic: Deprecate reboot type in GPREGRET

Deprecates setting GPREGRET to the reset reason as this has been
replaced with the boot mode retention subsystem for nRF51/nRF52.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 3.4: Add nRF51/nRF52 sys_reboot deprecation note

Adds a note that the old behaviour of setting the GPREGRET register
to the parameter of sys_reboot() is no longer the default as the
boot mode interface has replaced it.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: Add documentation for retention subsystem

Adds documentation for the retention subsystem and boot mode
interface. Also marks the retention subsystem as experimental.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
tests: Add mcuboot_recovery_retention test

Adds a test that exercises the boot mode (retention subsystem)
feature to check that an application can successfully boot into
mcuboot's serial recovery mode.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented on include/zephyr/retention/retention.h:
```diff
@@ -0,0 +1,113 @@
+
+typedef ssize_t (*data_retention_size_api)(const struct device *dev);
+typedef int (*data_retention_is_valid_api)(const struct device *dev);
+typedef int (*data_retention_read_api)(const struct device *dev, size_t offset, uint8_t *buffer,
+				       size_t size);
+typedef int (*data_retention_write_api)(const struct device *dev, size_t offset,
+					const uint8_t *buffer, size_t size);
```
change to off_t

nordicjm commented on include/zephyr/retention/retention.h:
```diff
@@ -0,0 +1,113 @@
+ *
+ * @retval		0 on success else negative errno code.
+ */
+int data_retention_read(const struct device *dev, size_t offset, uint8_t *buffer, size_t size);
+
+/**
+ * @brief		Writes data to the data retention area (underlying data does not need to be
+ *			cleared prior to writing), once function returns with a success code, the
+ *			data will be classed as valid if queried using data_retention_is_valid().
+ *
+ * @param dev		Data retention device to use.
+ * @param offset	Offset to write data to.
+ * @param buffer	Data to write.
+ * @param size		Size of data to be written.
+ *
+ * @retval		0 on success else negative errno code.
+ */
+int data_retention_write(const struct device *dev, size_t offset, const uint8_t *buffer,
```
as above

nordicjm commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,397 @@
+	return rc;
+}
+
+int data_retention_read(const struct device *dev, size_t offset, uint8_t *buffer, size_t size)
```
As above

nordicjm commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,397 @@
+	return rc;
+}
+
+int data_retention_write(const struct device *dev, size_t offset, const uint8_t *buffer,
```
As above

nordicjm commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,397 @@
+int data_retention_read(const struct device *dev, size_t offset, uint8_t *buffer, size_t size)
+{
+	const struct retention_config *config = dev->config;
+#ifdef CONFIG_MULTITHREADING
+	struct retention_data *data = dev->data;
+#endif
+	int rc;
+
+	if ((offset + size) > (config->size - config->reserved_size)) {
+		/* Disallow reading past the virtual data size */
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_MULTITHREADING
+	k_mutex_lock(&data->lock, K_FOREVER);
+#endif
+
+	rc = retained_mem_read(config->parent, (config->offset + config->prefix_len + offset),
+			       buffer, size);
+
+#ifdef CONFIG_MULTITHREADING
+	k_mutex_unlock(&data->lock);
+#endif
```
Move to lock() and unlock() functions for all file functions

nordicjm committed:
scripts: ci: check_compliance: Ignore mcuboot Kconfigs

Adds mcuboot Kconfig options to the ignore list as these are
defined in mcuboot but used in documentation and samples which
utilise sysbuild to build mcuboot.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

Laczen commented:
This whole system seems to make choices that are specific to mcuboot and as such should not be incorporated in the zephyr tree. 

TBH most of the mcuboot/mcumgr related stuff does not belong in the zephyr tree but should be part of the mcuboot/zephyr-mcuboot repo.

nordicjm commented:
> This whole system seems to make choices that are specific to mcuboot and as such should not be incorporated in the zephyr tree.

Like what? A shared memory region or register with a uint8 which can be read, written to or checked? How is that mcuboot specific? For your information the nRF5 SDK bootloader uses a similar method as does a third party's QSPI-based bootloader.

> TBH most of the mcuboot/mcumgr related stuff does not belong in the zephyr tree but should be part of the mcuboot/zephyr-mcuboot repo.

The mcuboot parts are part of mcuboot, as the above bot comment shows

nordicjm commented on dts/bindings/retention/zephyr,data-retention.yaml:
```diff
@@ -0,0 +1,37 @@
+      data size). Value is size in bytes (0 for none, 1 for 8-bit CRC, 2 for
+      16-bit CRC, 4 for 32-bit CRC).
+    type: int
+    default: 0
```
Fixed

nordicjm commented on dts/bindings/retention/zephyr,data-retention.yaml:
```diff
@@ -0,0 +1,37 @@
+  offset:
+    type: int
+    default: 0
+    description: Offset of area in parent node.
```
Fixed

nordicjm commented on dts/bindings/retention/zephyr,data-retention.yaml:
```diff
@@ -0,0 +1,37 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Data retention system area, which has a retained memory parent.
```
Added

Laczen commented:
As I said in the architecture meeting: I don't see why this should be added. What this PR adds is a specification on how to store data in the retention system and use devicetree (dts) as a means to communicate this to other applications (e.g. a bootloader).

**The devicetree (dts) should not be used for this**.

An application can directly use the retention system without this layer to achieve the same result.

Adding this PR will result in pollution of the dts files describing boards to add this functionality while its use will be limited to very specific cases.

mbolivar-nordic commented:
> The devicetree (dts) should not be used for this.

@laczen can you please clarify why you believe this? I read your comment, but it doesn't make an argument for why this is the case.

Laczen commented:
> > The devicetree (dts) should not be used for this.
> 
> @Laczen can you please clarify why you believe this? I read your comment, but it doesn't make an argument for why this is the case.

@mbolivar-nordic, if you look at an example from a dts:
```
      retention0: retention {
        compatible = "zephyr,data-retention";
        status = "okay";
        size = <64>;
        prefix = [04 fa];
        checksum = <1>;
      };
```
the "zephyr,data-retention" is describing the format of what is found in the retention area: a prefix, a checksum and the size (to determine the real data). This is a pure software choice. Why is this in the dts ? Because it matters for another application that is compiled with the same dts.

mbolivar-nordic commented:
> This is a pure software choice

This is permitted in some cases: https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#the-zephyr-prefix

This PR seems to be using the zephyr, prefix correctly in this case.

Am I missing something?

nordicjm commented on dts/bindings/retention/zephyr,data-retention.yaml:
```diff
@@ -0,0 +1,70 @@
+  uint8_t data_value;
+  int rc = data_retention_write(retention_dev, 0, &data_value,
+                                sizeof(data_value));
+
```
Got it, fixed.

Laczen commented:
> > This is a pure software choice
> 
> This is permitted in some cases: https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#the-zephyr-prefix
> 
> This PR seems to be using the zephyr, prefix correctly in this case.
> 
> Am I missing something?

The documentation states: `Therefore, as a compromise, we do allow some software configuration in Zephyr’s devicetree bindings, as long as they use this prefix to show that they are Zephyr specific.`, what is proposed in this PR is a complete software construct defined by devicetree bindings, a lot more than just some software configuration.

@nordicjm, could you please provide a example where the following retention data is used:
1. Data for a bootloader with magic "c0ffee", 4 bytes data and a 1 byte crc,
2. Data for the application with magic 0x61 0x70 0x70 0x64 ("appd"), 4 bytes and a 1 byte crc,
3. Data for the application with magic 0x6D 0x65 0x73 0x68 0x64 ("meshd"), 64 bytes and 2 byte crc,

nordicjm commented:
> @nordicjm, could you please provide a example where the following retention data is used:

No, there is text included already in the documentation describing how it works, how to use it and examples.

Laczen commented:
> @nordicjm, could you please provide a example where the following retention data is used:
> 
> 1. Data for a bootloader with magic "c0ffee", 4 bytes data and a 1 byte crc,
> 2. Data for the application with magic 0x61 0x70 0x70 0x64 ("appd"), 4 bytes and a 1 byte crc,
> 3. Data for the application with magic 0x6D 0x65 0x73 0x68 0x64 ("meshd"), 64 bytes and 2 byte crc,

This example would add the following to the dts:
```
sram@2003FFC0 {
    compatible = "zephyr,memory-region", "mmio-sram";
    reg = <0x2003FF80 128>;
    zephyr,memory-region = "RetainedMem";
    status = "okay";
    retainedmem {
      compatible = "zephyr,retained-ram";
      status = "okay";
      retention0: retention {
        compatible = "zephyr,data-retention";
        status = "okay";
        size = <11>;
        prefix = [0c 00 0f 0f 0e 0e];
        checksum = <1>;
      };
      retention1: retention {
        compatible = "zephyr,data-retention";
        status = "okay";
        size = <9>;
        prefix = [61 70 70 64];
        checksum = <1>;
      };
      retention2: retention {
        compatible = "zephyr,data-retention";
        status = "okay";
        size = <71>;
        prefix = [6D 65 73 68 64];
        checksum = <2>;
      };
    };
  };
```
And this is just a simple example...

nordicjm commented:
> > @nordicjm, could you please provide a example where the following retention data is used:
> > 
> > 1. Data for a bootloader with magic "c0ffee", 4 bytes data and a 1 byte crc,
> > 2. Data for the application with magic 0x61 0x70 0x70 0x64 ("appd"), 4 bytes and a 1 byte crc,
> > 3. Data for the application with magic 0x6D 0x65 0x73 0x68 0x64 ("meshd"), 64 bytes and 2 byte crc,
> 
> This example would add the following to the dts:
> 
> ```
> sram@2003FFC0 {
>     compatible = "zephyr,memory-region", "mmio-sram";
>     reg = <0x2003FF80 128>;
>     zephyr,memory-region = "RetainedMem";
>     status = "okay";
>     retainedmem {
>       compatible = "zephyr,retained-ram";
>       status = "okay";
>       retention0: retention {
>         compatible = "zephyr,data-retention";
>         status = "okay";
>         size = <11>;
>         prefix = [0c 00 0f 0f 0e 0e];
>         checksum = <1>;
>       };
>       retention1: retention {
>         compatible = "zephyr,data-retention";
>         status = "okay";
>         size = <9>;
>         prefix = [61 70 70 64];
>         checksum = <1>;
>       };
>       retention2: retention {
>         compatible = "zephyr,data-retention";
>         status = "okay";
>         size = <71>;
>         prefix = [6D 65 73 68 64];
>         checksum = <2>;
>       };
>     };
>   };
> ```
> 
> And this is just a simple example...

So?

Laczen commented:
> 
> So?

OK, if this is not bad enough explain what happens when applications want to use the retention area differently with different data (The "coffee" part is shared, as well as the "appd", but not the "meshd" part) ?

nordicjm commented:
> > So?
> 
> OK, if this is not bad enough explain what happens when applications want to use the retention area differently with different data (The "coffee" part is shared, as well as the "appd", but not the "meshd" part) ?

So like what happens if someone builds hello world, flashes it to the mcu, then builds blinky, and flashes is to the mcu, the application code area flash is changed? Users can do whatever they want in the retention areas, that's up to them, you use each section for the intended purpose, you don't reuse it for other things.

Laczen commented:
> > > So?
> > 
> > 
> > OK, if this is not bad enough explain what happens when applications want to use the retention area differently with different data (The "coffee" part is shared, as well as the "appd", but not the "meshd" part) ?
> 
> So like what happens if someone builds hello world, flashes it to the mcu, then builds blinky, and flashes is to the mcu, the application code area flash is changed? Users can do whatever they want in the retention areas, that's up to them, you use each section for the intended purpose, you don't reuse it for other things.

Each application that needs a change to the retention data will need to provide its own overlay file.

de-nordic reviewed (changes requested):
Some comments. I think it generally looks OK.
The MCUboot changes should not go with this PR.
I have used this PR to build MCUboot with just adding overlay:
```
/ {
        chosen {
                zephyr,boot-mode = &boot_mode0;
        };
};

&gpregret1 {
        status = "okay";

        boot_mode0: boot_mode {
                compatible = "zephyr,data-retention";
                status = "okay";
                size = <1>;
        };
};
```
and MCUBOOT_SERIAL enabled, and MCUboot stopped running at all, the without retention even enabled.

My test platform was nrf52840dk_nrf52840 and MCUboot and without this overlay the MCUboot works fine, including the serial recovery mode.

nordicjm commented:
> Some comments. I think it generally looks OK.
> The MCUboot changes should not go with this PR.
> I have used this PR to build MCUboot with just adding overlay:

You're not setting the boot partition, so that will place mcuboot at 0xc000 (if using nrf52)

de-nordic commented:
>
> An application can directly use the retention system without this layer to achieve the same result.

But the application needs to synchronize usage of retention API with other systems. Reserving space via DTS (overlays) allows to easily see how much retention using features can be used at once, depending on size of provided mem.

> 
> Adding this PR will result in pollution of the dts files describing boards to add this functionality while its use will be limited to very specific cases.

It does not seem that any DTS files get modified so far. I think it will be up to user to add all DTS entries for things they want to use with retention API.

nordicjm commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,401 @@
+	CHECKSUM_TYPE_NONE = 0,
+	CHECKSUM_TYPE_CRC8,
+	CHECKSUM_TYPE_CRC16,
+	CHECKSUM_TYPE_UNUSED,
```
Size is in bytes, there isn't a 3 byte checksum

de-nordic replied:
But still this `config->checksum == CHECKSUM_TYPE_CRC8` happens to figure out the size.

nordicjm commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,401 @@
+			goto finish;
+		}
+
+		if (config->checksum == CHECKSUM_TYPE_CRC8) {
+			uint8_t read_checksum;
+
+			rc = retained_mem_read(config->parent, (config->offset + data_size),
+					       (void *)&read_checksum, sizeof(read_checksum));
+			expected_checksum = (uint32_t)read_checksum;
+		} else if (config->checksum == CHECKSUM_TYPE_CRC16) {
+			uint16_t read_checksum;
+
+			rc = retained_mem_read(config->parent, (config->offset + data_size),
+					       (void *)&read_checksum, sizeof(read_checksum));
+			expected_checksum = (uint32_t)read_checksum;
+		} else if (config->checksum == CHECKSUM_TYPE_CRC32) {
+			rc = retained_mem_read(config->parent, (config->offset + data_size),
+					       (void *)&expected_checksum,
+					       sizeof(expected_checksum));
+		}
```
I was thinking of this way originally but will cause problems on big vs little endian

de-nordic replied:
Ah right, on big-endians, the
```
    rc = retained_mem_read(config->parent, (config->offset + data_size),
					       (void *)&read_checksum, size);
```
would have to look like this:
```
    rc = retained_mem_read(config->parent, (config->offset + data_size),
					       (void *)&read_checksum + sizeof(read_checksum) - size, size);
```

nordicjm commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,401 @@
+	const struct retention_config *config = dev->config;
+	int rc;
+
+	if (((size_t)offset + size) > (config->size - config->reserved_size)) {
```
Fixed

nordicjm commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,401 @@
+
+	data_retention_lock_take(dev);
+
+	if (((size_t)offset + size) > (config->size - config->reserved_size)) {
```
Fixed

nordicjm commented on tests/boot/mcuboot_recovery_retention/boards/nrf52840dk_nrf52840.overlay:
Fixed

Laczen commented:
> > An application can directly use the retention system without this layer to achieve the same result.
> 
> But the application needs to synchronize usage of retention API with other systems. Reserving space via DTS (overlays) allows to easily see how much retention using features can be used at once, depending on size of provided mem.
>
If it would be just reserving space, but this is describing the content in devicetree.

de-nordic commented:
> > Some comments. I think it generally looks OK.
> > The MCUboot changes should not go with this PR.
> > I have used this PR to build MCUboot with just adding overlay:
> 
> You're not setting the boot partition, so that will place mcuboot at 0xc000 (if using nrf52)

Right. Fixed and it works fine!

de-nordic reviewed (dismissed):
Look OK.
I have tested the MCUboot changes, with retention and they seem to work as expected.
There are few Copyrights missing, and I am not in favour some `ifs` (https://github.com/zephyrproject-rtos/zephyr/pull/55125#discussion_r1151631833), but generally looks fine.
Tested on nrf52840dk_nrf52840 with MCUboot from the manifest of this PR and seems to work fine.

mbolivar-nordic commented:
@Laczen I'm having trouble following what your objection is. At first it sounded like you didn't want this in devicetree at all. Now it sounds like you're OK with reserving memory area base addresses and sizes in devicetree, but not their contents. Is that right?

If so, is it a personal preference, or is there a technical problem?

ldenefle commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,401 @@
+	uint8_t buffer[CONFIG_RETENTION_BUFFER_SIZE];
+	off_t pos;
+
+	data_retention_lock_take(dev);
```
quick question, are the locking mechanisms really necessary containing the underlying subsys `retained_mem` has its own synchronization primitives ?

nordicjm replied:
If there are 2 threads using this system and there are locks on the retained memory but not on the retention subsystem, one thread can call a write function which starts writing data to the retained memory, say an interrupt occurs and now the other thread writes to the same retention area - what will the data look like once both functions have returned? I do take your point though that there should be a way to disable using mutexes if the user so wishes not to (at their own risk) so can add a Kconfig to both retained memory and retention to disable mutex support

ldenefle replied:
ok gotcha, I didn't figure out that the `data_retention_write` was making several calls to `retained_mem_write` in that case  you could have partial writes which would be annoying error cases to handle.

ldenefle commented:
It  would have been easier to review / merge if everything related to bootmode was in a subsequent PR, anyway, is there a reason why it's the `data_retention_X` api and not the `retention_X` api  ? It seems more instinctive to call it the latter as all the source / headers are called `retention.{h,c}`

nordicjm commented:
> It would have been easier to review / merge if everything related to bootmode was in a subsequent PR, anyway, is there a reason why it's the `data_retention_X` api and not the `retention_X` api ? It seems more instinctive to call it the latter as all the source / headers are called `retention.{h,c}`

Retention is a common enough word/name that I'm concerned that it might cause name collisions with external libraries or code, we currently have that issue with some other code in the kernel and want to avoid it, but can rename to just retention if people think it won't be an issue

nordicjm commented:
It's a good question though, I'll tag for dev review later and ask people's opinions

nordicjm commented:
No objections or issues raised at dev review meeting to having name as just retention, if there are issues in future, will have to resolve it then, will update PR with new name when mcuboot PR is merged

Laczen commented:
> @Laczen I'm having trouble following what your objection is. At first it sounded like you didn't want this in devicetree at all. Now it sounds like you're OK with reserving memory area base addresses and sizes in devicetree, but not their contents. Is that right?
> 
> If so, is it a personal preference, or is there a technical problem?

@mbolivar-nordic, to be clear: I don't want this in devicetree at all:
a. This uses devicetree as a means to communicate a software configuration,
b. It is not needed at all, there is a retention area that can be used to share information. The subsystem that used it needs to agree on what the format is.
c. It pollutes the dts files,

So it is neither a personal preference or a technical problem, this is about how to use the supplied tools.

Laczen commented:
> @Laczen I am making a good faith effort to understand, but I'm failing to see your point.
> 
> > @mbolivar-nordic, to be clear: I don't want this in devicetree at all:
> > a. This uses devicetree as a means to communicate a software configuration,
> 
> You and I have established already in previous comments that there is precedent and policy for doing software configuration in devicetree in Zephyr.
> 
> This pull request complies with the written policy in the devicetree documentation.
> 
> Therefore, unless there is a change in policy, I do not consider this a valid objection to the PR from a devicetree perspective.

Is it OK to translate this as extensions to devicetree will be accepted when they follow the policy. No discussion about this.

> > b. It is not needed at all, there is a retention area that can be
> > used to share information. The subsystem that uses it needs to agree
> > on what the format is.
> 
> It seems clear to me that there is a need to share retained memory between different applications, which implies a need to configure the areas where the retained data goes. So I assume you're not saying there is no need for retained data _at all_ -- but instead the objection is about how and where it should be configured.
> 

Please, in zephyr we define a nvs system that can contain whatever data by just pointing to a partition. This could also be done for retained data, just point to the retained memory area where it can be found.

BTW, It should be clear for everyone that this PR is not about reserving an area for shared data, this PR describes how the data in this area should look like.

> > c. It pollutes the dts files,
> 
> "Pollute" in this context is a pejorative and emotional/aesthetic word choice. This is not introducing a toxic chemical or trash into an environment.

Maybe I should have used different phrasing (english is not my mother tongue), I am afraid that boards dts files will be "flooded" (is this ok?) with definitions as soon as we accept uses of devicetree like in the proposed PR.

Having said this I am stopping the discussion. I won't approve and think it is a bad proposal.

thedjnK commented:
> Please, in zephyr we define a nvs system that can contain whatever data by just pointing to a partition. This could also be done for retained data, just point to the retained memory area where it can be found.

Right, so no different to this

> BTW, It should be clear for everyone that this PR is not about reserving an area for shared data, this PR describes how the data in this area should look like.

No it does not

> Maybe I should have used different phrasing (english is not my mother tongue), I am afraid that boards dts files will be "flooded" (is this ok?) with definitions as soon as we accept uses of devicetree like in the proposed PR.

Why would it become flooded? This is mostly a downstream addition, I don't envisage many in-tree boards to use this by default

> Having said this I am stopping the discussion. I won't approve and think it is a bad proposal.

Great!

nordicjm committed:
drivers: retained_mem: Add optional mutex disablement

Adds an optional Kconfig to disable use of mutexes in the retained
memory driver.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
retention: Add optional mutex disablement

Adds an optional Kconfig to disable use of mutexes in the
retention subsystem.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

wbober commented on doc/services/retention/index.rst:
```diff
@@ -0,0 +1,199 @@
+should it be stored to a non-volatile backing store such as :ref:`flash_api`,
+:ref:`eeprom_api` or battery-backed RAM.
+
+The retention system builds on top of the retained data driver, and adds
```
Please link to docs that describe `retained data driver`.

wbober replied:
What I meant was to put the link in the rst.

wbober replied:
Is my understanding correct that the subsystem is only might not necessarily be compatible with any retained mem driver? For example, this would not work with https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/retained_mem/retained_mem_nrf_gpregret.c, correct?

nordicjm replied:
It would work fine with it. You can even see the test here https://github.com/zephyrproject-rtos/zephyr/pull/55125/commits/d7d180b17baf480e3cfd426a98e643acecc88dfd which runs on nRF52

wbober commented on doc/services/retention/index.rst:
```diff
@@ -0,0 +1,199 @@
+*****************
+
+To use the retention system, a retained data driver must be setup for the board
+you are using, there is a zephyr driver which can be used which will use some
```
Which zephyr driver? 

From my understanding of the code this must be `retained_mem_zephyr_ram.c` driver. Correct? If that is correct, please state this.

nordicjm replied:
No, any retained memory driver.

FrancescoSer commented on doc/releases/release-notes-3.4.rst:
```diff
@@ -123,6 +123,16 @@ Deprecated in this release
   board-specific configuration in board Kconfig fragments in the ``boards``
   folder of the application.
 
+* For nRF51/nRF52-based boards, the behaviour of reset reason provided
+  to :c:func:`sys_reboot` being set in the GPREGRET register has been dropped,
+  this function will now just reboot the device without changing the register
+  contents. The new method for setting this register is by using the boot mode
+  feature of the retention subsystem, see the
+  :ref:`boot mode API <boot_mode_api>` for details. To restore the old
+  deprecated functionality of reset reason of :c:func:`sys_reboot` being set to
+  the GPREGRET register, enable
+  :kconfig:option:`CONFIG_NRF_STORE_REBOOT_TYPE_GPREGRET`.
```
```suggestion
* On nRF51 and nRF52-based boards, the behaviour of the reset reason provided
  to :c:func:`sys_reboot` and being set in the GPREGRET register has been dropped.
  This function will now just reboot the device without changing the register
  content. The new method for setting this register uses the boot mode
  feature of the retention subsystem, see the
  :ref:`boot mode API <boot_mode_api>` for details. To restore the
  deprecated functionality, enable
  :kconfig:option:`CONFIG_NRF_STORE_REBOOT_TYPE_GPREGRET`.
```

FrancescoSer commented on doc/services/retention/index.rst:
```diff
@@ -0,0 +1,199 @@
+Retention System
+################
+
+The retention system provides an API which allows applications to read/write
+data from/to memory areas or devices which retain the data whilst the device is
+powered. This is designed for sharing information between different
+applications or for a single application when a device reboots without losing
+state information, the contents of the data should not persist in the event of
+a power failure (or possibly with some low power modes on some devices) nor
+should it be stored to a non-volatile backing store such as :ref:`flash_api`,
+:ref:`eeprom_api` or battery-backed RAM.
```
```suggestion
The retention system provides an API which allows applications to read and write
data from and to memory areas or devices that retain the data while the device is
powered. This allows for sharing information between different
applications or within a single application without losing
state information when a device reboots .
The stored data should not persist in the event of
a power failure (or during some low-power modes on some devices) nor
should it be stored to a non-volatile storage like :ref:`flash_api`,
:ref:`eeprom_api`, or battery-backed RAM.
```

FrancescoSer reviewed (changes requested):
Just some suggestions

Laczen commented:
@nordicjm, could you show how separate subsystems (e.g. mcuboot, coredebug) can define data-retention area's independently from each other (so without knowledge of the other size or offset).

thedjnK commented:
> @nordicjm, could you show how separate subsystems (e.g. mcuboot, coredebug) can define data-retention area's independently from each other (so without knowledge of the other size or offset).

Did you actually look at the PR?

Laczen commented:
> > @nordicjm, could you show how separate subsystems (e.g. mcuboot, coredebug) can define data-retention area's independently from each other (so without knowledge of the other size or offset).
> 
> Did you actually look at the PR?

Maybe I am overlooking something, can I just create them in separate overlay files?

thedjnK commented:
> > > @nordicjm, could you show how separate subsystems (e.g. mcuboot, coredebug) can define data-retention area's independently from each other (so without knowledge of the other size or offset).
> > 
> > 
> > Did you actually look at the PR?
> 
> Maybe I am overlooking something, can I just create them in separate overlay files?

Yes, same as flash partitions, can define a base board with a large retained memory area then add retention partitions from overlays at specific offsets and of specified length

Laczen commented:
> > > > @nordicjm, could you show how separate subsystems (e.g. mcuboot, coredebug) can define data-retention area's independently from each other (so without knowledge of the other size or offset).
> > > 
> > > 
> > > Did you actually look at the PR?
> > 
> > 
> > Maybe I am overlooking something, can I just create them in separate overlay files?
> 
> Yes, same as flash partitions, can define a base board with a large retained memory area then add retention partitions from overlays at specific offsets and of specified length

Flash partitions are not independent, they are coupled by offset and size. To avoid making errors they are best placed in one dts file.

nordicjm committed:
doc: release: 3.4: Add retention subsystem and boot mode entries

Adds details on the newly introduced retention subsystem and boot
mode APIs to the release notes.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

ldenefle commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,400 @@
+	size_t offset;
+	size_t size;
+	size_t reserved_size;
+	uint8_t checksum;
```
Should this parameter be renamed `checksum_len` or `checksum_size` it would make some code reading a bit easier like `ssize_t data_size = config->size - config->checksum;`

nordicjm replied:
Fixed

ldenefle commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,400 @@
+	retention_lock_take(dev);
+
+	if (offset < 0 || ((size_t)offset + size) > (config->size - config->reserved_size)) {
+		/* Disallow reading past the virtual data size or before it */
```
Need to change in comment `writing` not `reading`

nordicjm replied:
Fixed

ldenefle commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,400 @@
+			goto finish;
+		}
+
+		pos += sizeof(buffer);
```
Shouldn't this be 
```
pos += MIN((config->size - pos), sizeof(buffer));
```
?

ldenefle replied:
I don't think it should cause a bug though

nordicjm replied:
Fixed

ldenefle reviewed (commented):
I just have some style comments

ldenefle commented on subsys/retention/retention.c:
```diff
@@ -0,0 +1,400 @@
+			rc = retained_mem_read(config->parent, pos, buffer, read_size);
+
+			if (rc < 0) {
+				goto finish;
```
I understand you use `goto` in most of the `retention` implementation but I get the idea that you mostly do it to enforce proper release of mutex

nordicjm replied:
goto is used throughout the file to do this alongside drivers and other subsystems so want to keep it the same

ldenefle replied:
ok fair

ldenefle commented on subsys/retention/bootmode.c:
```diff
@@ -0,0 +1,44 @@
+
+	if (rc == 1) {
+		uint8_t stored_mode;
+		int rc = retention_read(boot_mode_dev, 0, &stored_mode, sizeof(stored_mode));
```
There's a check missing here for if `rc < 0`

nordicjm replied:
If rc is less then 0 then it will be returned directly to the caller, though I did miss removing the second rc definition

ldenefle replied:
ok maybe I'm missing something, bear with me
```
	int rc;

	rc = retention_is_valid(boot_mode_dev);

	if (rc == 1) {
		uint8_t stored_mode;

		rc = retention_read(boot_mode_dev, 0, &stored_mode, sizeof(stored_mode));

		if (stored_mode == boot_mode) {
			rc = 1;
		} else {
			rc = 0;
		}
	}
```
on `retention_read` if `rc < 0` I'm guessing `stored_mode != boot_mode` which will then set `rc = 0` so for the caller of `bootmode_check` the return is 0 but actually there was an issue with  `retention_read` that was not returned.

Do you agree ?

My suggested fix would be to just check for `rc > 0` then implement the logic ternary-equivalent to `rc = stored_mode == boot_mode ? 1 : 0;` (not saying you should use a ternary)

nordicjm replied:
Ah I see what you mean, you are correct, should be fixed now.

ldenefle commented:
I have one more comment but otherwise looks good to go to me

Laczen commented:
The proposed solution seems flawed: the retention system defines an area in a retention memory that includes a optional header and crc check to enable validating that the "retained data" is valid. However as both these are optional (and they need to be because it also supports only 1 byte of info) the validity check is not OK when header and crc are disabled. In that case the retention system should return `-ENOTSUP` (or something equivalent).
The validity check should thus be performed otherwise. The whole construct of header and crc check can than just as well be integrated in the code that uses the retention memory.

nordicjm commented:
> The proposed solution seems flawed: the retention system defines an area in a retention memory that includes a optional header and crc check to enable validating that the "retained data" is valid. However as both these are optional (and they need to be because it also supports only 1 byte of info) the validity check is not OK when header and crc are disabled. In that case the retention system should return -ENOTSUP (or something equivalent).

Thanks for spotting, fixed

Laczen commented:
> > The proposed solution seems flawed: the retention system defines an area in a retention memory that includes a optional header and crc check to enable validating that the "retained data" is valid. However as both these are optional (and they need to be because it also supports only 1 byte of info) the validity check is not OK when header and crc are disabled. In that case the retention system should return -ENOTSUP (or something equivalent).
> 
> Thanks for spotting, fixed

@nordicjm, the proposed fix is making `-ENOTSUP` a valid check for the boot_mode retention area, this is a strange choice.

It seems that the first subsystem that is making use of the data retention subsystem makes everything defined in the data retention system unnecessary, so what is the need to add the prefix and the crc ? If there are other subsystems that could benefit from it please show how several subsystems should set up the data retention system and show how they can be enabled and used independently and if needed what changes are required to use them independently (app using boot_mode, app using ..._mode, app using boot_mode + ..._mode).

BTW: the data retention system is using `offset` and `size` properties in the compatible, these should be combined in a `reg` property.

nordicjm committed:
dts: arm: nordic: Add address and size cell values to GPREGRET

Adds address cells of size 1 and size cells of size 1 to GPREGRET
instances for Nordic devices.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
dts: bindings: retained_mem: Add address and size cell values

Adds address and size cell constant values of 1 each.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
> I successfully ran the tests on a DK, would it be worth for me to add some overlays with different retentions backends configurations? If I do this would you consider adding those commits to the PR ?

Once this is merged, sure, submit a PR to add support for more boards.

nordicjm commented:
> BTW: the data retention system is using `offset` and `size` properties in the compatible, these should be combined in a `reg` property.

Fixed

nordicjm commented:
> > > BTW: the data retention system is using `offset` and `size` properties in the compatible, these should be combined in a `reg` property.
> > 
> > 
> > Fixed
> 
> Good, still missing the sample to show how multiple subsystems (independently) can enable/disable this functionality according to their needs. If this is not possible or there are limitations it is important to understand this in the evaluation of the proposed PR.

No different than flash partitions

Laczen commented:
> > > > BTW: the data retention system is using `offset` and `size` properties in the compatible, these should be combined in a `reg` property.
> > > 
> > > 
> > > Fixed
> > 
> > 
> > Good, still missing the sample to show how multiple subsystems (independently) can enable/disable this functionality according to their needs. If this is not possible or there are limitations it is important to understand this in the evaluation of the proposed PR.
> 
> No different than flash partitions

Flash partitions have a lot of limitations. There are afaik no subsystems that are making changes (e.g. inserting) to the flash partitions because it is error prone and not possible without prior knowledge. The only way that changes are applied is by deleting the complete partition layout and rebuilding it. 
This is not modular at all and cannot be used by independent modules trying to use this.

So please again, show how to do it.
